File: week1/week1/python/imageDraw.py
# Copyright 2017 BIG VISION LLC ALL RIGHTS RESERVED
# 
# This code is made available to the students of 
# the online course titled "Computer Vision for Faces" 
# by Satya Mallick for personal non-commercial use. 
#
# Sharing this code is strictly prohibited without written
# permission from Big Vision LLC. 
#
# For licensing and other inquiries, please email 
# spmallick@bigvisionllc.com 
#

import cv2
import numpy as np

image = np.zeros((400, 400, 3), dtype=np.uint8)

# Draw a line
# We are making a copy of image because we don't want to draw all the shapes on one image
imageLine = image.copy()
cv2.line(imageLine, (15, 20), (370, 350), (110, 220, 0), thickness=2, lineType=cv2.LINE_AA)
cv2.imshow("line", imageLine)

# Draw a circle
imageCircle = image.copy()
cv2.circle(imageCircle, (200, 200), 50, (0, 0, 255), thickness=2, lineType=cv2.LINE_AA)
cv2.imshow("circle", imageCircle)

# Draw an ellipse
# IMP Note: Ellipse Centers and Axis lengths must be integers
imageEllipse = image.copy()
cv2.ellipse(imageEllipse, (200, 200), (100, 160), 45, 0, 360, (255, 0, 0), thickness=2, lineType=cv2.LINE_AA)
cv2.ellipse(imageEllipse, (200, 200), (100, 160), 135, 0, 360, (0, 0, 255), thickness=5, lineType=cv2.LINE_AA)
cv2.ellipse(imageEllipse, (200, 200), (150, 50), 135, 0, 360, (0, 255, 0), thickness=2, lineType=cv2.LINE_AA)
cv2.imshow("ellipse", imageEllipse)

# Draw a rectangle (thickness is a positive integer)
imageRectangle = image.copy()
cv2.rectangle(imageRectangle, (105, 105), (205, 205), (0, 55, 255), thickness=+2, lineType=cv2.LINE_8)
cv2.imshow("rectangle1", imageRectangle)
# Draw a filled rectangle (thickness is a negative integer)
cv2.rectangle(imageRectangle, (210, 210), (310, 310), (100, 155, 25), thickness=-1, lineType=cv2.LINE_8)
cv2.imshow("rectangle2", imageRectangle)

# Draw a filled Polygon
imagePolygon = image.copy()
w = 400
# Create some points
rookPoints = np.zeros((1, 20, 2))
rookPoints[0, 0, :] = (w/4.0, 7*w/8.0)
rookPoints[0, 1, :] = (3*w/4.0, 7*w/8.0)
rookPoints[0, 2, :] = (3*w/4.0, 13*w/16.0)
rookPoints[0, 3, :] = (11*w/16.0, 13*w/16.0)
rookPoints[0, 4, :] = (19*w/32.0, 3*w/8.0)
rookPoints[0, 5, :] = (3*w/4.0, 3*w/8.0)
rookPoints[0, 6, :] = (3*w/4.0, w/8.0)
rookPoints[0, 7, :] = (26*w/40.0, w/8.0)
rookPoints[0, 8, :] = (26*w/40.0, w/4.0)
rookPoints[0, 9, :] = (22*w/40.0, w/4.0)
rookPoints[0, 10, :] = (22*w/40.0, w/8.0)
rookPoints[0, 11, :] = (18*w/40.0, w/8.0)
rookPoints[0, 12, :] = (18*w/40.0, w/4.0)
rookPoints[0, 13, :] = (14*w/40.0, w/4.0)
rookPoints[0, 14, :] = (14*w/40.0, w/8.0)
rookPoints[0, 15, :] = (w/4.0, w/8.0)
rookPoints[0, 16, :] = (w/4.0, 3*w/8.0)
rookPoints[0, 17, :] = (13*w/32.0, 3*w/8.0)
rookPoints[0, 18, :] = (5*w/16.0, 13*w/16.0)
rookPoints[0, 19, :] = (w/4.0, 13*w/16.0)

# fillPoly expects data points to be in int32 format
ppt = np.int32(np.around(rookPoints))

#Fill polygon formed by the points
cv2.fillPoly(imagePolygon, ppt, (255, 255, 255), lineType=cv2.LINE_8)
cv2.imshow("Image", imagePolygon)

# Put text into image
imageText = image.copy()
cv2.putText(imageText, "OpenCV Basics", (100, 180), cv2.FONT_HERSHEY_SIMPLEX, 1, (0, 200, 100), 4)
cv2.imshow("text", imageText)

cv2.waitKey(0)

cv2.destroyAllWindows()

File: week1/week1/cpp/imageDraw.cpp
/*
Copyright 2017 BIG VISION LLC ALL RIGHTS RESERVED

This code is made available to the students of 
the online course titled "Computer Vision for Faces" 
by Satya Mallick for personal non-commercial use. 

Sharing this code is strictly prohibited without written
permission from Big Vision LLC. 

For licensing and other inquiries, please email 
spmallick@bigvisionllc.com 
*/

#include <opencv2/core.hpp>
#include <opencv2/imgproc.hpp>
#include <opencv2/highgui.hpp>

using namespace cv;
 
int main(void) {
    // Create black empty images
    Mat image = Mat::zeros(400, 400, CV_8UC3);

    // Draw a line
    // We are making a copy of image because we don't want to draw all the shapes on one image
    Mat imageLine = image.clone();
    line(imageLine, Point(15, 20), Point(370, 350), Scalar(110, 220, 0), 5, CV_AA);
    imshow("line", imageLine);

    // Draw a circle
    Mat imageCircle = image.clone();
    circle(imageCircle, Point(200, 200), 50.0, Scalar(0, 0, 255), 2, 8);
    imshow("circle", imageCircle);

    // Draw an ellipse
    Mat imageEllipse = image.clone();
    ellipse(imageEllipse, Point(200, 200), Size(100.0, 160.0), 45, 0, 360, Scalar(255, 0, 0), 1, 8);
    ellipse(imageEllipse, Point(200, 200), Size(100.0, 160.0), 135, 0, 360, Scalar(255, 0, 0), 10, 8);
    ellipse(imageEllipse, Point(200, 200), Size(150.0, 50.0), 135, 0, 360, Scalar(0, 255, 0), 1, 8);
    imshow("ellipse", imageEllipse);

    // Draw a rectangle (5th argument is not -ve)
    Mat imageRectangle = image.clone();
    rectangle(imageRectangle, Point(15, 20), Point(115, 120), Scalar(0, 55, 255), +1, 4 );
    imshow("rectangle1", imageRectangle);
    // Draw a filled rectangle (5th argument is -ve)
    rectangle(imageRectangle, Point(125, 135), Point(225, 235), Scalar(100, 155, 25), -1, 8);
    imshow("rectangle2", imageRectangle);

    // Draw a filled Polygon
    Mat imagePolygon = image.clone();
    int w = 400;
    /** Create some points */
    Point rookPoints[1][20];
    rookPoints[0][0] = Point(w/4.0, 7*w/8.0);
    rookPoints[0][1] = Point(3*w/4.0, 7*w/8.0);
    rookPoints[0][2] = Point(3*w/4.0, 13*w/16.0);
    rookPoints[0][3] = Point(11*w/16.0, 13*w/16.0);
    rookPoints[0][4] = Point(19*w/32.0, 3*w/8.0);
    rookPoints[0][5] = Point(3*w/4.0, 3*w/8.0);
    rookPoints[0][6] = Point(3*w/4.0, w/8.0);
    rookPoints[0][7] = Point(26*w/40.0, w/8.0);
    rookPoints[0][8] = Point(26*w/40.0, w/4.0);
    rookPoints[0][9] = Point(22*w/40.0, w/4.0);
    rookPoints[0][10] = Point(22*w/40.0, w/8.0);
    rookPoints[0][11] = Point(18*w/40.0, w/8.0);
    rookPoints[0][12] = Point(18*w/40.0, w/4.0);
    rookPoints[0][13] = Point(14*w/40.0, w/4.0);
    rookPoints[0][14] = Point(14*w/40.0, w/8.0);
    rookPoints[0][15] = Point(w/4.0, w/8.0);
    rookPoints[0][16] = Point(w/4.0, 3*w/8.0);
    rookPoints[0][17] = Point(13*w/32.0, 3*w/8.0);
    rookPoints[0][18] = Point(5*w/16.0, 13*w/16.0);
    rookPoints[0][19] = Point(w/4.0, 13*w/16.0);

    const Point* ppt[1] = {rookPoints[0]};
    int npt[] = {20};
    //Fill the polygon formed by the points 
    fillPoly(imagePolygon, ppt, npt, 1, Scalar(255, 255, 255), 8);
    imshow("Image", imagePolygon);

    // Put text into image
    Mat imageText = image.clone();
    putText(imageText, "OpenCV Basics", Point(100,180), FONT_HERSHEY_SIMPLEX, 1, Scalar(0, 200, 200), 4);
    imshow("text", imageText);

    waitKey(0);
    return 0;
}